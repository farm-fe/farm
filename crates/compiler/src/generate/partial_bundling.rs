use std::sync::Arc;

use farmfe_core::{
  config::partial_bundling::PartialBundlingEnforceResourceConfig,
  context::CompilationContext,
  error::CompilationError,
  hashbrown::HashSet,
  module::{module_graph::ModuleGraph, module_group::ModuleGroupGraph, Module, ModuleId},
  plugin::PluginHookContext,
  resource::{
    resource_pot::{ResourcePot, ResourcePotType},
    resource_pot_map::ResourcePotMap,
  },
};

pub fn partial_bundling(
  context: &Arc<CompilationContext>,
  hook_context: &PluginHookContext,
) -> farmfe_core::error::Result<()> {
  #[cfg(feature = "profile")]
  farmfe_core::puffin::profile_function!();

  let module_group_graph = analyze_module_graph(context, hook_context)?;
  // insert the module group map into the context
  let mut context_module_group_graph = context.module_group_graph.write();
  context_module_group_graph.replace(module_group_graph);
  drop(context_module_group_graph);
  let resource_pot_map = generate_resource_pot_map(context, hook_context)?;
  // insert the resource pot graph into the context
  let mut g = context.resource_pot_map.write();
  g.replace(resource_pot_map);

  Ok(())
}

fn analyze_module_graph(
  context: &Arc<CompilationContext>,
  hook_context: &PluginHookContext,
) -> farmfe_core::error::Result<ModuleGroupGraph> {
  let mut module_graph = context.module_graph.write();

  let module_group_graph = context
    .plugin_driver
    .analyze_module_graph(&mut module_graph, context, hook_context)?
    .ok_or(CompilationError::PluginHookResultCheckError {
      hook_name: "analyze_module_graph".to_string(),
    })?;

  Ok(module_group_graph)
}

pub fn generate_resource_pot_map(
  context: &Arc<CompilationContext>,
  hook_context: &PluginHookContext,
) -> farmfe_core::error::Result<ResourcePotMap> {
  let (enforce_resource_pots, modules) = generate_enforce_resource_pots(context);

  let mut resources_pots =
    call_partial_bundling_hook(&modules.into_iter().collect(), context, hook_context)?;
  // extends enforce resource pots
  resources_pots.extend(enforce_resource_pots);
  fill_necessary_fields_for_resource_pot(resources_pots.iter_mut().collect(), context);

  let mut resource_pot_map = ResourcePotMap::new();

  for resource_pot in resources_pots {
    resource_pot_map.add_resource_pot(resource_pot);
  }

  Ok(resource_pot_map)
}

pub fn get_enforce_resource_name_for_module(
  module_id: &ModuleId,
  enforce_resources: &Vec<PartialBundlingEnforceResourceConfig>,
) -> Option<String> {
  for enforce_resource_config in enforce_resources {
    if enforce_resource_config
      .test
      .iter()
      .any(|test| test.is_match(&module_id.to_string()))
    {
      return Some(enforce_resource_config.name.clone());
    }
  }

  None
}

pub fn call_partial_bundling_hook(
  modules: &Vec<ModuleId>,
  context: &Arc<CompilationContext>,
  hook_context: &PluginHookContext,
) -> farmfe_core::error::Result<Vec<ResourcePot>> {
  let res = context
    .plugin_driver
    .partial_bundling(modules, context, hook_context)?
    .ok_or(CompilationError::PluginHookResultCheckError {
      hook_name: "partial_bundling".to_string(),
    })?;

  Ok(res)
}

pub fn fill_necessary_fields_for_resource_pot(
  resources_pots: Vec<&mut ResourcePot>,
  context: &Arc<CompilationContext>,
) {
  let mut module_graph = context.module_graph.write();
  let mut module_group_graph = context.module_group_graph.write();

  for resource_pot in resources_pots {
    if resource_pot.module_groups.is_empty() {
      let mut module_groups = HashSet::new();
      let mut entry_module = None;
      for module_id in resource_pot.modules() {
        let module = module_graph.module_mut(module_id).unwrap();
        module.resource_pot.push(resource_pot.id.clone());
        module_groups.extend(module.module_groups.clone());

        if module_graph.entries.contains_key(module_id) {
          if entry_module.is_some() {
            panic!("a resource pot can only have one entry module");
          }
          entry_module = Some(module_id.clone());
        }
      }

      resource_pot.entry_module = entry_module;
      resource_pot.module_groups = module_groups;
    }

    for module_group_id in &resource_pot.module_groups {
      let module_group = module_group_graph
        .module_group_mut(module_group_id)
        .unwrap();
      module_group.add_resource_pot(resource_pot.id.clone());
    }
  }
}

pub fn get_resource_pot_id_for_enforce_resources(
  name: String,
  module_id: &ModuleId,
  module_graph: &ModuleGraph,
) -> (ResourcePotType, String, String) {
  let module = module_graph.module(module_id).unwrap();
  let resource_pot_type = ResourcePotType::from(module.module_type.clone());
  let id = ResourcePot::gen_id(&name, resource_pot_type.clone());

  (resource_pot_type, name, id)
}

pub fn get_resource_pot_id_for_enforce_resources_by_removed_module(
  name: String,
  module: &Module,
) -> (ResourcePotType, String, String) {
  let resource_pot_type = ResourcePotType::from(module.module_type.clone());
  let id = ResourcePot::gen_id(&name, resource_pot_type.clone());

  (resource_pot_type, name, id)
}

fn generate_enforce_resource_pots(
  context: &Arc<CompilationContext>,
) -> (Vec<ResourcePot>, Vec<ModuleId>) {
  let mut modules = HashSet::new();
  let mut enforce_resource_pot_map = ResourcePotMap::new();
  let module_graph = context.module_graph.read();
  let module_group_graph = context.module_group_graph.read();

  // generate enforce resource pots first
  for g in module_group_graph.module_groups() {
    for module_id in g.modules() {
      // ignore external module
      if module_graph.module(module_id).unwrap().external {
        continue;
      }

      if let Some(name) = get_enforce_resource_name_for_module(
        module_id,
        &context.config.partial_bundling.enforce_resources,
      ) {
        let (resource_pot_type, resource_pot_name, resource_pot_id) =
          get_resource_pot_id_for_enforce_resources(name.clone(), module_id, &module_graph);

        let resource_pot = enforce_resource_pot_map.resource_pot_mut(&resource_pot_id);

        if let Some(resource_pot) = resource_pot {
          resource_pot.add_module(module_id.clone());
        } else {
          let mut resource_pot = ResourcePot::new(resource_pot_name, resource_pot_type);
          resource_pot.add_module(module_id.clone());
          enforce_resource_pot_map.add_resource_pot(resource_pot);
        }
      } else {
        // if the module is not in any enforce resource pot, add it modules for partial bundling
        modules.insert(module_id.clone());
      }
    }
  }

  (
    enforce_resource_pot_map.take_resource_pots(),
    modules.into_iter().collect(),
  )
}

#[cfg(test)]
mod tests {
  use std::sync::Arc;

  use farmfe_core::{
    config::{
      config_regex::ConfigRegex, partial_bundling::PartialBundlingEnforceResourceConfig, Config,
    },
    context::CompilationContext,
    hashbrown::HashSet,
    plugin::{Plugin, PluginHookContext},
  };
  use farmfe_plugin_partial_bundling::module_group_graph_from_entries;
  use farmfe_testing_helpers::construct_test_module_graph_complex;

  use super::generate_resource_pot_map;

  #[test]
  fn test_generate_resource_pot_map() {
    let mut module_graph = construct_test_module_graph_complex();
    let module_group_graph = module_group_graph_from_entries(
      &module_graph
        .entries
        .clone()
        .into_iter()
        .map(|(entry, _)| entry)
        .collect(),
      &mut module_graph,
    );

    let mut config = Config::default();
    config.partial_bundling.enforce_resources = vec![PartialBundlingEnforceResourceConfig {
      name: "test".into(),
      test: vec![ConfigRegex::new("F"), ConfigRegex::new("H")],
    }];
    let plugins: Vec<Arc<dyn Plugin + 'static>> = vec![Arc::new(
      farmfe_plugin_partial_bundling::FarmPluginPartialBundling::new(&config),
    )];
    let context = Arc::new(CompilationContext::new(config, plugins).unwrap());

    {
      let mut mg = context.module_graph.write();
      *mg = module_graph;
    }

    {
      let mut mgg = context.module_group_graph.write();
      *mgg = module_group_graph;
    }

    let resource_pot_map =
      generate_resource_pot_map(&context, &PluginHookContext::default()).unwrap();

    let mut resource_pots = resource_pot_map.take_resource_pots();
    resource_pots.sort_by_key(|p| p.id.clone());

    assert_eq!(resource_pots.len(), 5);
    // A, C
    assert_eq!(resource_pots[0].modules(), vec![&"A".into(), &"C".into()]);
    assert_eq!(resource_pots[0].entry_module, Some("A".into()));
    // B, E
    assert_eq!(resource_pots[1].modules(), vec![&"B".into(), &"E".into()]);
    assert_eq!(resource_pots[1].entry_module, Some("B".into()));
    // D
    assert_eq!(resource_pots[2].modules(), vec![&"D".into()]);
    // G
    assert_eq!(resource_pots[3].modules(), vec![&"G".into()]);
    // F, H
    assert_eq!(resource_pots[4].modules(), vec![&"F".into(), &"H".into()]);

    // assert necessary fields are filled
    assert_eq!(
      resource_pots[0].module_groups,
      HashSet::from(["A".into(), "F".into()])
    );
    assert_eq!(resource_pots[0].entry_module, Some("A".into()));

    assert_eq!(resource_pots[1].module_groups, HashSet::from(["B".into()]));
    assert_eq!(resource_pots[1].entry_module, Some("B".into()));

    assert_eq!(
      resource_pots[2].module_groups,
      HashSet::from(["D".into(), "B".into()])
    );
    assert_eq!(resource_pots[2].entry_module, None);

    assert_eq!(resource_pots[3].module_groups, HashSet::from(["G".into()]));
    assert_eq!(resource_pots[3].entry_module, None);

    assert_eq!(
      resource_pots[4].module_groups,
      HashSet::from(["F".into(), "G".into(), "B".into(), "D".into()])
    );
    assert_eq!(resource_pots[4].entry_module, None);
  }

  #[test]
  fn test_generate_resource_pot_map_external() {
    let mut module_graph = construct_test_module_graph_complex();
    // mark H as external
    module_graph.module_mut(&"H".into()).unwrap().external = true;

    let module_group_graph = module_group_graph_from_entries(
      &module_graph
        .entries
        .clone()
        .into_iter()
        .map(|(entry, _)| entry)
        .collect(),
      &mut module_graph,
    );

    let config = Config::default();
    let plugins: Vec<Arc<dyn Plugin + 'static>> = vec![Arc::new(
      farmfe_plugin_partial_bundling::FarmPluginPartialBundling::new(&config),
    )];
    let context = Arc::new(CompilationContext::new(config, plugins).unwrap());

    {
      let mut mg = context.module_graph.write();
      *mg = module_graph;
    }

    {
      let mut mgg = context.module_group_graph.write();
      *mgg = module_group_graph;
    }

    let resource_pot_map =
      generate_resource_pot_map(&context, &PluginHookContext::default()).unwrap();

    let mut resource_pots = resource_pot_map.take_resource_pots();
    resource_pots.sort_by_key(|p| p.id.clone());

    assert_eq!(resource_pots.len(), 5);
    // A, C
    assert_eq!(resource_pots[0].modules(), vec![&"A".into(), &"C".into()]);
    assert_eq!(resource_pots[0].entry_module, Some("A".into()));
    // B, E
    assert_eq!(resource_pots[1].modules(), vec![&"B".into(), &"E".into()]);
    assert_eq!(resource_pots[1].entry_module, Some("B".into()));
    // D
    assert_eq!(resource_pots[2].modules(), vec![&"D".into()]);
    // F, H
    assert_eq!(resource_pots[3].modules(), vec![&"F".into()]);
    // G
    assert_eq!(resource_pots[4].modules(), vec![&"G".into()]);

    // assert necessary fields are filled
    assert_eq!(
      resource_pots[0].module_groups,
      HashSet::from(["A".into(), "F".into()])
    );
    assert_eq!(resource_pots[0].entry_module, Some("A".into()));

    assert_eq!(resource_pots[1].module_groups, HashSet::from(["B".into()]));
    assert_eq!(resource_pots[1].entry_module, Some("B".into()));

    assert_eq!(
      resource_pots[2].module_groups,
      HashSet::from(["D".into(), "B".into()])
    );
    assert_eq!(resource_pots[2].entry_module, None);

    assert_eq!(resource_pots[3].module_groups, HashSet::from(["F".into()]));
    assert_eq!(resource_pots[3].entry_module, None);

    assert_eq!(resource_pots[4].module_groups, HashSet::from(["G".into()]));
    assert_eq!(resource_pots[4].entry_module, None);
  }
}
