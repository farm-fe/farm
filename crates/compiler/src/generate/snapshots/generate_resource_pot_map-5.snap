---
source: crates/compiler/src/generate/partial_bundling.rs
input_file: crates/compiler/src/generate/partial_bundling.rs
---
(
    (
        "immutable",
        false,
    ),
    (
        "modules",
        [
            ModuleId {
                relative_path: "F",
                query_string: "",
            },
            ModuleId {
                relative_path: "H",
                query_string: "",
            },
        ],
    ),
    (
        "resource_pot_type",
        Custom(
            "__farm_unknown",
        ),
    ),
    (
        "name",
        "test",
    ),
    (
        "id",
        "test_custom(\"__farm_unknown\")",
    ),
    (
        "entry",
        None,
    ),
    (
        "resources",
        [],
    ),
    (
        "module_groups",
        [
            ModuleId {
                relative_path: "B",
                query_string: "",
            },
            ModuleId {
                relative_path: "D",
                query_string: "",
            },
            ModuleId {
                relative_path: "F",
                query_string: "",
            },
            ModuleId {
                relative_path: "G",
                query_string: "",
            },
        ],
    ),
)
